### PR TITLE
Publish Docker images + Refactor GH Action

### PR DESCRIPTION
**Pull request by ktham**
TL:DR Lots of cleanup on how tests run + actually publish images to Docker Hub!

## Changes
* Refactor GH actions and remove `test-backend.sh`
* **Conditionally run** `backend` tests if backend code has changed or if it's `master` branch
* (TODO for future) **Conditionally run** `frontend` tests if frontend code has changed or if it's `master` branch
* Build docker images
* Add simple **integration/end-to-end test** to validate docker images
  - Check for `pong` response on `http://backend:8000/ping/`
  - Check for `HTTP 200` on `http://frontend:3000/`
  - Check for `HTTP 200` on `load_balancer`
  - Check for `"<title>ISMP: International Student Mentorship Program</title>"` on `localhost:8080` (`load_balancer`)
* Fix docker-compose mounting: don't override `node_modules` directory from image
* **Publish** docker images on `master` commits
* Clean up `Makefile` and remove unnecessary make tasks
* Add comments in various areas for clarity

## Types of changes

- [x] I've labeled this PR with appropriate labels (`comp/*` component, bug/feature, documentation and other less important ones).

## Checklist
_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. This is simply a reminder of what we are going to look for before merging your code._

- [x] I've linked the proper Github Issue(s) to my PR
- [x] Lint and unit tests pass locally with my changes
- [x] I have added tests (unit, integration...etc.) that prove my fix is effective or that my change works
- [x] I have added or updated necessary documentation (if appropriate). You can search, for instance using the [`silver searcher`](https://github.com/ggreer/the_silver_searcher) `ag <STRING>` to find all usages in the repository.
- [x] I have notified the team of any breaking or important changes

### CODE REVIEW

File: .github/workflows/main.yml [hunk 1]
```diff
@@ -11,10 +11,75 @@ on:
...
+          fi
+          if git diff --quiet HEAD master -- frontend ; then
+            echo "::set-output name=frontend::yes"
+          else
+            echo "::set-output name=frontend::no"
+          fi
+
+      - run: docker-compose build
+
+      - name: 'Run Backend Tests'
```
Comment (ktham) [inline .github/workflows/main.yml:41]: This is where we conditionally run backend tests

File: .github/workflows/main.yml [hunk 2]
```diff
@@ -11,10 +11,75 @@ on:
...
+        if: steps.changes.outputs.backend == 'yes'
+        run: docker-compose -f ./docker-compose.yml -f ./docker-compose.test.yml run --rm test_backend
+
+      - name: 'Run Frontend Tests'
```
Comment (ktham) [inline .github/workflows/main.yml:45]: This is where we _would_ conditionally run frontend tests, but it's disabled for now

File: .github/workflows/main.yml [hunk 3]
```diff
@@ -11,10 +11,75 @@ on:
...
+        # TODO:` `yarn run lint` does not work
+        # TODO: `yarn test does` not run any tests
+        if: false # steps.changes.outputs.frontend == 'yes'
+        run: docker-compose -f ./docker-compose.yml -f ./docker-compose.test.yml run --rm test_frontend
+
+      - name: 'Run Simple end-to-end Tests on Docker Images'
+        run: |
+          docker-compose up -d
+          docker-compose run --rm load_balancer curl --max-time 5 --retry 10 --retry-connrefused backend:8000/ping/ | grep pong
```
Comment (ktham) [inline .github/workflows/main.yml:54]: Here, we have a simple end-to-end test that validates whether our docker images play well together!

File: .github/workflows/main.yml [hunk 4]
```diff
@@ -11,10 +11,75 @@ on:
...
+        if: (github.ref == 'refs/heads/master') || (steps.changes.outputs.backend == 'yes')
+        run: docker-compose -f ./docker-compose.yml -f ./docker-compose.test.yml run --rm test_backend
+
+      - name: 'Run Frontend Tests'
+        # TODO:` `yarn run lint` does not work
+        # TODO: `yarn test does` not run any tests
+        if: false # (github.ref == 'refs/heads/master') || (steps.changes.outputs.frontend == 'yes')
+        run: docker-compose -f ./docker-compose.yml -f ./docker-compose.test.yml run --rm test_frontend
+
+      - name: 'Run Simple end-to-end Tests on Docker Images'
```
Comment (ktham) [inline .github/workflows/main.yml:51]: These are also the steps to run the website locally 😄 it's now part of an automated test

Comment (douglaschan32167) [reply .github/workflows/main.yml:51]: nice! Thanks Kevin!

File: docker-compose.test.yml [hunk 5]
```diff
@@ -21,28 +24,16 @@ services:
...
-      - test_backend
-
+  # Run frontend tests
   test_frontend:
     image: "localhost/ismp/frontend:snapshot"
+    build: ./frontend
     environment:
-      REACT_APP_API_HOST: "http://test_backend:8000"
-      USE_S3: "FALSE"
+      # TODO: Remove CI=true after https://github.com/facebook/create-react-app/pull/8845
```
Comment (douglaschan32167) [inline docker-compose.test.yml:35]: It looks like the PR linked here is merged. Should this be removed now?

File: docker-compose.yml [hunk 6]
```diff
@@ -58,16 +57,17 @@ services:
...
 
   frontend:
-    # An image called "localhost/ismp/frontend:snapshot" will be built using the build context in ./frontend
     image: "localhost/ismp/frontend:snapshot"
     build: ./frontend
-    stdin_open: true
-    tty: true
     environment:
       REACT_APP_API_HOST: "http://localhost:8080"
+      # TODO: Remove CI=true after https://github.com/facebook/create-react-app/pull/8845
```
Comment (douglaschan32167) [inline docker-compose.yml:68]: See comment above

File: docker-compose.test.yml (hunk 5 above)
Comment (ktham) [reply docker-compose.test.yml:35]: yeah, it's merged, but seems like it hasn't made it to a release yet
